hibernate-core/src/test/java/org/hibernate/orm/test/inheritance/discriminator/JoinedInheritanceEagerTest.java

```diff
@@ -0,0 +1,238 @@
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Test cases for joined inheritance with eager fetching.
+ *
+ * @author Christian Beikov
+ */
+@DomainModel(
+		annotatedClasses = {
+				JoinedInheritanceEagerTest.BaseEntity.class,
+				JoinedInheritanceEagerTest.EntityA.class,
+				JoinedInheritanceEagerTest.EntityB.class,
+				JoinedInheritanceEagerTest.EntityC.class,
+				JoinedInheritanceEagerTest.EntityD.class
+		}
+)
+@SessionFactory
+public class JoinedInheritanceEagerTest {
+
+	@BeforeEach
+	public void setUp(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			EntityC entityC = new EntityC( 1L );
+			EntityD entityD = new EntityD( 2L );
+
+			EntityB entityB = new EntityB( 3L );
+			entityB.setRelation( entityD );
+
+			EntityA entityA = new EntityA( 4L );
+			entityA.setRelation( entityC );
+
+			session.persist( entityC );
+			session.persist( entityD );
+			session.persist( entityA );
+			session.persist( entityB );
+		} );
+	}
+
+	@AfterEach
+	public void cleanUp(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					EntityA entityA = session.get( EntityA.class, 4L );
+					EntityB entityB = session.get( EntityB.class, 3L );
+					EntityD entityD = session.get( EntityD.class, 2L );
+					EntityC entityC = session.get( EntityC.class, 1L );
+
+					session.delete( entityD );
+					session.delete( entityC );
+					session.delete( entityA );
+					session.delete( entityB );
+				}
+		);
+	}
+
+	@Test
+	@TestForIssue(jiraKey = "HHH-12375")
+	public void joinFindEntity(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			EntityA entityA = session.get( EntityA.class, 4L );
+			assertTrue( Hibernate.isInitialized( entityA.getRelation() ) );
+			assertFalse( Hibernate.isInitialized( entityA.getAttributes() ) );
+		} );
+	}
+
+	@Test
+	@TestForIssue(jiraKey = "HHH-12375")
+	public void joinFindParenEntity(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			BaseEntity baseEntity = session.get( BaseEntity.class, 4L );
+			assertThat( baseEntity, notNullValue() );
+			assertThat( baseEntity, instanceOf( EntityA.class ) );
+			assertTrue( Hibernate.isInitialized( ( (EntityA) baseEntity ).getRelation() ) );
+			assertFalse( Hibernate.isInitialized( ( (EntityA) baseEntity ).getAttributes() ) );
+		} );
+
+		scope.inTransaction( session -> {
+			BaseEntity baseEntity = session.get( BaseEntity.class, 3L );
+			assertThat( baseEntity, notNullValue() );
+			assertThat( baseEntity, instanceOf( EntityB.class ) );
+			assertTrue( Hibernate.isInitialized( ( (EntityB) baseEntity ).getRelation() ) );
+			assertFalse( Hibernate.isInitialized( ( (EntityB) baseEntity ).getAttributes() ) );
+		} );
+	}
+
+	@Test
+	@TestForIssue(jiraKey = "HHH-12375")
+	public void selectBaseType(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			List result = session.createQuery( "from BaseEntity" ).list();
+			assertEquals( result.size(), 2 );
```

Comment:
Correct usage is `assertEquals( 2, result.size())`, but consistent usage of `assertThat` as following would be better:
```
assertThat( result, hasSize( 2 ));
```